### ISSUE TITLE
Avoid repeatedly constructing identical proxy objects.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8238]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8238|
|Type|enhancement|
|Created|2016-03-18 19:17:32Z|
|Branch|https://github.com/twisted/twisted/tree/static-proxy-object-8238|

As pointed out by fijal on IRC, creating new class objects every time we make a web request is bonkers slow, particularly on PyPy which is essentially entirely unable to optimise the operation away.

Given that this class is identical every time we do this, we may as well just create it once and then use the same class every time.

There are no functional changes here.

Attachments:

* [8238_1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a3b/a3b43377db732fd82b6179549e7d06d67c75e8a5/70591a799068e15d2aaf6b09c0cf55b29f88977d.patch) (873 bytes) - added by Lukasa on 2016-03-18 19:19:17Z - First draft patch

<details><summary>Searchable metadata</summary>

```
trac-id__8238 8238
type__enhancement enhancement
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__branches_static_proxy_object_8238 branches/static-proxy-object-8238
branch_author__mithrandi mithrandi
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1458328652426688 1458328652426688
changetime__1474198049187875 1474198049187875
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

The attached patch hoists the class invocation out to module scope.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_2"></a>|@mithrandi commented|
|-|-|

(In [47026]) Branching to static-proxy-object-8238.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_3"></a>|@mithrandi commented|
|-|-|

Running the "web" benchmark from twisted-benchmarks, I get these results:

CPython 2.7.11, Twisted trunk@47013: 1172.0 web/sec (5860 web in 5 seconds)

CPython 2.7.11, this branch: 1384.0 web/sec (6920 web in 5 seconds)

PyPy 5.0.0, Twisted trunk@47013: 754.0 web/sec (3770 web in 5 seconds)

PyPy 5.0.0, this branch: 1336.0 web/sec (6680 web in 5 seconds)

So that seems like a pretty clear win.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_4"></a>|@mithrandi set status to `closed`|
|-|-|

(In [47029]) Merge static-proxy-object-8238: Create proxyForInterface(IClientRequest) type statically in twisted.web.client.

Author: lukasa
Reviewer: mithrandi
Fixes: [#8238](https://github.com/twisted/twisted/issues/8238)

This improves performance of twisted-benchmarks/web on CPython by 18%,
and by 77% on PyPy.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
